### PR TITLE
Removed dependency on 'org.apache.commons.io'

### PR DESCRIPTION
**Pull request by cweitkamp**
- Removed dependency on `org.apache.commons.io`

- [ ] bom/compile/pom.xml
- [ ] bom/compile/pom.xml
- [ ] features/karaf/openhab-tp/src/main/feature/feature.xml
- [x] Resolve `itest.bndrun`

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

File: bundles/org.openhab.core.id/src/main/java/org/openhab/core/id/InstanceUUID.java [hunk 1]
```diff
@@ -75,19 +73,21 @@ public static synchronized String get() {
...
+            LOGGER.warn("Failed writing to file '{}': {}", content, file.getAbsolutePath(), ioe.getMessage());
         }
     }
 
     private static String readFirstLine(File file) {
-        List<String> lines = null;
         try {
-            lines = IOUtils.readLines(new FileInputStream(file));
+            final List<String> lines = Files.readAllLines(file.toPath(), StandardCharsets.UTF_8);
+            return lines.isEmpty() ? "" : lines.get(0);
```
Comment (cweitkamp) [inline bundles/org.openhab.core.id/src/main/java/org/openhab/core/id/InstanceUUID.java:86]: We only need the first line. Could be changed to:

```suggestion
            final BufferedReader reader = Files.newBufferedReader(path);
            return reader.readLine();
```

File: bundles/org.openhab.core.io.rest.auth/src/main/java/org/openhab/core/io/rest/auth/internal/JwtHelper.java [hunk 2]
```diff
@@ -78,13 +77,14 @@ private RsaJsonWebKey generateNewKey() throws JoseException, FileNotFoundExcepti
...
+        Files.writeString(file.toPath(), keyJson, StandardCharsets.UTF_8);
         return newKey;
     }
 
     private RsaJsonWebKey loadOrGenerateKey() throws FileNotFoundException, JoseException, IOException {
         try {
-            List<String> lines = IOUtils.readLines(new FileInputStream(KEY_FILE_PATH));
+            File file = new File(KEY_FILE_PATH);
+            List<String> lines = Files.readAllLines(file.toPath(), StandardCharsets.UTF_8);
             return (RsaJsonWebKey) JsonWebKey.Factory.newJwk(lines.get(0));
```
Comment (cweitkamp) [inline bundles/org.openhab.core.io.rest.auth/src/main/java/org/openhab/core/io/rest/auth/internal/JwtHelper.java:88]: We only need the first line. Could be changed to:

```suggestion
            final BufferedReader reader = Files.newBufferedReader(file.toPath());
            return (RsaJsonWebKey) JsonWebKey.Factory.newJwk(reader.readLine());
```

File: bundles/org.openhab.core.config.dispatch/src/main/java/org/openhab/core/config/dispatch/internal/ConfigDispatcher.java [hunk 3]
```diff
@@ -300,7 +300,7 @@ private void internalProcessConfigFile(File configFile) throws IOException, File
         String context = null;
 
         // configuration file contains a PID Marker
-        List<String> lines = IOUtils.readLines(new FileInputStream(configFile));
+        List<String> lines = Files.readAllLines(configFile.toPath(), StandardCharsets.UTF_8);
         String exclusivePID = lines.size() > 0 ? getPIDFromLine(lines.get(0)) : null;
```
Comment (cweitkamp) [inline bundles/org.openhab.core.config.dispatch/src/main/java/org/openhab/core/config/dispatch/internal/ConfigDispatcher.java:304]:
```suggestion
        final BufferedReader reader = Files.newBufferedReader(configFile.toPath());
        String exclusivePID = getPIDFromLine(reader.readLine());
```

File: itests/org.openhab.core.config.dispatch.tests/src/main/java/org/openhab/core/config/dispatch/internal/ConfigDispatcherOSGiTest.java [hunk 4]
```diff
@@ -1028,22 +1029,22 @@ private void verifyNoPropertiesForConfiguration(String pid) {
...
             return null;
         }
-        try (FileInputStream fileInputStream = new FileInputStream(file)) {
-            List<String> lines = IOUtils.readLines(fileInputStream);
-            for (String line : lines) {
+        // This method will return null, if there are no files in the directory.
+        final String separator = property + "=";
+        String value = null;
+        try {
+            for (String line : Files.readAllLines(file.toPath(), StandardCharsets.UTF_8)) {
```
Comment (cweitkamp) [inline itests/org.openhab.core.config.dispatch.tests/src/main/java/org/openhab/core/config/dispatch/internal/ConfigDispatcherOSGiTest.java:1047]: Could be changed to `Files.lines(file.toPath()).foreach(...)`

File: bundles/org.openhab.core.id/src/main/java/org/openhab/core/id/InstanceUUID.java [hunk 5]
```diff
@@ -75,19 +73,21 @@ public static synchronized String get() {
     private static void writeFile(File file, String content) throws IOException {
         // create intermediary directories
         file.getParentFile().mkdirs();
-        try (OutputStream outputStream = new FileOutputStream(file)) {
-            IOUtils.write(content, outputStream);
+        try {
+            Files.writeString(file.toPath(), content, StandardCharsets.UTF_8);
+        } catch (IOException ioe) {
+            LOGGER.warn("Failed writing to file '{}': {}", content, file.getAbsolutePath(), ioe.getMessage());
```
Comment (cweitkamp) [inline bundles/org.openhab.core.id/src/main/java/org/openhab/core/id/InstanceUUID.java:79]:
```suggestion
            LOGGER.warn("Failed writing to file '{}': {}", file.getAbsolutePath(), ioe.getMessage());
```

File: bundles/org.openhab.core.config.dispatch/src/main/java/org/openhab/core/config/dispatch/internal/ConfigDispatcher.java (hunk 3 above)
Comment (Legion2) [reply bundles/org.openhab.core.config.dispatch/src/main/java/org/openhab/core/config/dispatch/internal/ConfigDispatcher.java:304]: all lines are required later (line 359)

Comment (cweitkamp) [reply bundles/org.openhab.core.config.dispatch/src/main/java/org/openhab/core/config/dispatch/internal/ConfigDispatcher.java:304]: Correct. Thanks for the hint.

File: bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/URLAudioStream.java [hunk 6]
```diff
@@ -65,24 +66,20 @@ private InputStream createInputStream() throws AudioException {
         try {
             switch (extension) {
                 case M3U_EXTENSION:
-                    try (final InputStream isM3U = new URL(url).openStream()) {
-                        for (final String line : IOUtils.readLines(isM3U)) {
-                            if (!line.isEmpty() && !line.startsWith("#")) {
-                                url = line;
-                                break;
-                            }
+                    for (final String line : Files.readAllLines(new File(url).toPath())) {
```
Comment (cpmeister) [inline bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/URLAudioStream.java:69]: Converting urls to files can cause a lot of problems if the urls don't have the right schema so create the path from a URI instead.
```suggestion
                    for (final String line : Files.readAllLines(Paths.get(URI.create(url)))) {
```

File: bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/URLAudioStream.java [hunk 7]
```diff
@@ -65,24 +66,20 @@ private InputStream createInputStream() throws AudioException {
...
                 case PLS_EXTENSION:
-                    try (final InputStream isPLS = new URL(url).openStream()) {
-                        for (final String line : IOUtils.readLines(isPLS)) {
-                            if (!line.isEmpty() && line.startsWith("File")) {
-                                final Matcher matcher = PLS_STREAM_PATTERN.matcher(line);
-                                if (matcher.find()) {
-                                    url = matcher.group(1);
-                                    break;
-                                }
+                    for (final String line : Files.readAllLines(new File(url).toPath())) {
```
Comment (cpmeister) [inline bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/URLAudioStream.java:77]:
```suggestion
                    for (final String line : Files.readAllLines(Paths.get(URI.create(url)))) {
```

File: bundles/org.openhab.core.model.core/src/main/java/org/openhab/core/model/core/internal/ModelRepositoryImpl.java [hunk 8]
```diff
@@ -98,7 +98,9 @@ public boolean addOrRefreshModel(String name, final InputStream originalInputStr
         try {
             InputStream inputStream = null;
             if (originalInputStream != null) {
-                byte[] bytes = IOUtils.toByteArray(originalInputStream);
+                final ByteArrayOutputStream output = new ByteArrayOutputStream();
+                originalInputStream.transferTo(output);
+                byte[] bytes = output.toByteArray();
```
Comment (cpmeister) [inline bundles/org.openhab.core.model.core/src/main/java/org/openhab/core/model/core/internal/ModelRepositoryImpl.java:103]: Does`originalInputStream.readAllBytes()` not work?

File: bundles/org.openhab.core.model.script/src/org/openhab/core/model/script/actions/HTTP.java [hunk 9]
```diff
@@ -147,16 +148,19 @@ static public String sendHttpPutRequest(String url, String contentType, String c
...
      */
-    static public String sendHttpPutRequest(String url, String contentType, String content, Map<String, String> headers, int timeout) {
+    static public String sendHttpPutRequest(String url, String contentType, String content, Map<String, String> headers,
+            int timeout) {
         try {
             Properties headerProperties = new Properties();
             headerProperties.putAll(headers);
-            return HttpUtil.executeUrl(HttpMethod.PUT.name(), url, headerProperties, IOUtils.toInputStream(content), contentType, timeout);
+            return HttpUtil.executeUrl(HttpMethod.PUT.name(), url, headerProperties,
+                    new ByteArrayInputStream(content.getBytes()), contentType, timeout);
```
Comment (cpmeister) [inline bundles/org.openhab.core.model.script/src/org/openhab/core/model/script/actions/HTTP.java:163]: Charset should be specified

File: bundles/org.openhab.core.model.script/src/org/openhab/core/model/script/actions/HTTP.java [hunk 10]
```diff
@@ -216,7 +220,8 @@ static public String sendHttpPostRequest(String url, String contentType, String
     static public String sendHttpPostRequest(String url, String contentType, String content, int timeout) {
         String response = null;
         try {
-            response = HttpUtil.executeUrl(HttpMethod.POST.name(), url, IOUtils.toInputStream(content), contentType, timeout);
+            response = HttpUtil.executeUrl(HttpMethod.POST.name(), url, new ByteArrayInputStream(content.getBytes()),
```
Comment (cpmeister) [inline bundles/org.openhab.core.model.script/src/org/openhab/core/model/script/actions/HTTP.java:223]: Charset should be specified

File: bundles/org.openhab.core.model.script/src/org/openhab/core/model/script/actions/HTTP.java [hunk 11]
```diff
@@ -225,19 +230,22 @@ static public String sendHttpPostRequest(String url, String contentType, String
...
      */
-    public static String sendHttpPostRequest(String url, String contentType, String content, Map<String, String> headers, int timeout) {
+    public static String sendHttpPostRequest(String url, String contentType, String content,
+            Map<String, String> headers, int timeout) {
         try {
             Properties headerProperties = new Properties();
             headerProperties.putAll(headers);
-            return HttpUtil.executeUrl(HttpMethod.POST.name(), url, headerProperties, IOUtils.toInputStream(content), contentType, timeout);
+            return HttpUtil.executeUrl(HttpMethod.POST.name(), url, headerProperties,
+                    new ByteArrayInputStream(content.getBytes()), contentType, timeout);
```
Comment (cpmeister) [inline bundles/org.openhab.core.model.script/src/org/openhab/core/model/script/actions/HTTP.java:248]: see above

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/firmware/FirmwareImpl.java [hunk 12]
```diff
@@ -176,7 +176,9 @@ public FirmwareRestriction getFirmwareRestriction() {
                 MessageDigest md = MessageDigest.getInstance("MD5");
 
                 try (DigestInputStream dis = new DigestInputStream(inputStream, md)) {
-                    bytes = IOUtils.toByteArray(dis);
+                    final ByteArrayOutputStream output = new ByteArrayOutputStream();
+                    dis.transferTo(output);
+                    bytes = output.toByteArray();
```
Comment (cpmeister) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/firmware/FirmwareImpl.java:181]: Same question as earlier

File: bundles/org.openhab.core.model.script/src/org/openhab/core/model/script/actions/HTTP.java [hunk 13]
```diff
@@ -135,7 +135,8 @@ static public String sendHttpPutRequest(String url, String contentType, String c
     static public String sendHttpPutRequest(String url, String contentType, String content, int timeout) {
         String response = null;
         try {
-            response = HttpUtil.executeUrl(HttpMethod.PUT.name(), url, IOUtils.toInputStream(content), contentType, timeout);
+            response = HttpUtil.executeUrl(HttpMethod.PUT.name(), url, new ByteArrayInputStream(content.getBytes()),
```
Comment (cpmeister) [inline bundles/org.openhab.core.model.script/src/org/openhab/core/model/script/actions/HTTP.java:138]: Charset should be specified

**Comment by cweitkamp** [comment]
@cpmeister Thanks for the useful review. I rebased and incorporated all suggested changes.

File: bundles/org.openhab.core.io.rest.auth/src/main/java/org/openhab/core/io/rest/auth/internal/JwtHelper.java [hunk 14]
```diff
@@ -78,14 +79,14 @@ private RsaJsonWebKey generateNewKey() throws JoseException, FileNotFoundExcepti
...
-        IOUtils.write(keyJson, new FileOutputStream(file));
+        Files.writeString(file.toPath(), keyJson, StandardCharsets.UTF_8);
         return newKey;
     }
 
     private RsaJsonWebKey loadOrGenerateKey() throws FileNotFoundException, JoseException, IOException {
         try {
-            List<String> lines = IOUtils.readLines(new FileInputStream(KEY_FILE_PATH));
-            return (RsaJsonWebKey) JsonWebKey.Factory.newJwk(lines.get(0));
+            final BufferedReader reader = Files.newBufferedReader(Paths.get(KEY_FILE_PATH));
```
Comment (cpmeister) [inline bundles/org.openhab.core.io.rest.auth/src/main/java/org/openhab/core/io/rest/auth/internal/JwtHelper.java:88]: You should also close this reader after you are done with it

**Comment by cweitkamp** [comment]
@wborn I resolved all conflicts. Unfortunately I had to squash my commits to simplify the rebase.

**Pull request merged**